datasource_test.go

```diff
@@ -224,6 +224,60 @@ func Test_error_retries(t *testing.T) {
 
 }
 
+func Test_error_source(t *testing.T) {
```

Comment:
Seems like a lot of repetitive bolierplate in this and other test functions, maybe makes sense to move it out of the test function and reuse to simplify reading the test itself?

Reply:
true.  let me try to clean this up